File: python/stt/mel.py
from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np

from .constants import (
    DEFAULT_N_MELS,
    WHISPER_HOP_LENGTH,
    WHISPER_NB_FRAMES,
    WHISPER_NFFT,
    WHISPER_PAD_SAMPLES,
    WHISPER_SAMPLING_HZ,
)

N_FRAMES = WHISPER_NB_FRAMES
N_FFT = WHISPER_NFFT
HOP_LENGTH = WHISPER_HOP_LENGTH
SAMPLE_RATE = WHISPER_SAMPLING_HZ
TARGET_SAMPLES = WHISPER_PAD_SAMPLES

_MEL_F_SP = 200.0 / 3.0
_MEL_MIN_LOG_HZ = 1_000.0

def _mel_min_log_mel() -> float:
    return _MEL_MIN_LOG_HZ / _MEL_F_SP

def _mel_logstep() -> float:
    return math.log(6.4) / 27.0

def hz_to_mel(f: float) -> float:
    if f >= _MEL_MIN_LOG_HZ:
        return _mel_min_log_mel() + math.log(f / _MEL_MIN_LOG_HZ) / _mel_logstep()
    return f / _MEL_F_SP

def mel_to_hz(m: float) -> float:
    if m >= _mel_min_log_mel():
        return _MEL_MIN_LOG_HZ * math.exp((m - _mel_min_log_mel()) * _mel_logstep())
    return _MEL_F_SP * m

def pad_or_truncate_to_30s(audio: np.ndarray) -> np.ndarray:
    if audio.dtype != np.float32:
        audio = audio.astype(np.float32)
    out = np.zeros(TARGET_SAMPLES, dtype=np.float32)
    n = min(len(audio), TARGET_SAMPLES)
    out[:n] = audio[:n]
    return out

def build_hann_window() -> np.ndarray:
    i = np.arange(N_FFT, dtype=np.float64)
    w = 0.5 - 0.5 * np.cos(2.0 * np.pi * i / float(N_FFT))
    return w.astype(np.float32)

def build_mel_filters(n_mels: int, n_fft: int = N_FFT, sample_rate: int = SAMPLE_RATE) -> np.ndarray:
    n_bins = n_fft // 2 + 1
    f_min = 0.0
    f_max = float(sample_rate) / 2.0
    m_min = hz_to_mel(f_min)
    m_max = hz_to_mel(f_max)

    mel_points = np.zeros(n_mels + 2, dtype=np.float64)
    for i in range(n_mels + 2):
        frac = float(i) / (float(n_mels) + 1.0)
        mel_points[i] = m_min + (m_max - m_min) * frac
    hz_points = np.array([mel_to_hz(float(m)) for m in mel_points], dtype=np.float64)

    fft_freqs = np.arange(n_bins, dtype=np.float64) * float(sample_rate) / float(n_fft)

    filters = np.zeros((n_mels, n_bins), dtype=np.float32)
    eps = float(np.finfo(np.float32).eps)
    for m in range(n_mels):
        lower = float(hz_points[m])
        center = float(hz_points[m + 1])
        upper = float(hz_points[m + 2])
        lower_slope = max(center - lower, eps)
        upper_slope = max(upper - center, eps)
        enorm = 2.0 / max(upper - lower, eps)
        for k in range(n_bins):
            freq = float(fft_freqs[k])
            weight = 0.0
            if lower <= freq <= center:
                weight = (freq - lower) / lower_slope
            elif center < freq <= upper:
                weight = (upper - freq) / upper_slope
            filters[m, k] = float(weight * enorm)
    return filters

@dataclass
class MelExtractor:
    n_mels: int = DEFAULT_N_MELS

    def __post_init__(self) -> None:
        self._filters = build_mel_filters(self.n_mels)
        self._hann = build_hann_window()

    @property
    def filters(self) -> np.ndarray:
        return self._filters

    @property
    def hann(self) -> np.ndarray:
        return self._hann

    def log_mel(self, audio_30s: np.ndarray) -> np.ndarray:
        if audio_30s.dtype != np.float32:
            audio_30s = audio_30s.astype(np.float32)
        if len(audio_30s) != TARGET_SAMPLES:
            raise ValueError(f"expected {TARGET_SAMPLES} samples (30s @ 16kHz), got {len(audio_30s)}")

        pad = N_FFT // 2
        padded = np.zeros(len(audio_30s) + N_FFT, dtype=np.float32)
        for i in range(pad):
            padded[i] = audio_30s[pad - i]
        padded[pad:pad + len(audio_30s)] = audio_30s
        for i in range(pad):
            src = max(len(audio_30s) - 2 - i, 0)
            padded[pad + len(audio_30s) + i] = audio_30s[src]

        n_bins = N_FFT // 2 + 1
        frame_starts = np.arange(N_FRAMES) * HOP_LENGTH
        frames = np.empty((N_FRAMES, N_FFT), dtype=np.float32)
        for i, start in enumerate(frame_starts):
            frames[i] = padded[start:start + N_FFT] * self._hann
        spec = np.fft.rfft(frames, n=N_FFT, axis=1)
        power = (spec.real * spec.real + spec.imag * spec.imag).astype(np.float32)

        mel = self._filters @ power.T

        eps = 1e-10
        mel = np.maximum(mel, eps)
        mel = np.log10(mel)
        max_val = float(mel.max())
        floor = max_val - 8.0
        mel = np.maximum(mel, floor)
        mel = (mel + 4.0) / 4.0
        return mel.astype(np.float32)

    def extract(self, samples: np.ndarray) -> np.ndarray:
        return self.log_mel(pad_or_truncate_to_30s(samples))

def log_mel_spectrogram(audio: np.ndarray, n_mels: int = DEFAULT_N_MELS) -> np.ndarray:
    return MelExtractor(n_mels=n_mels).extract(audio)

File: rust/src/stt/mel.rs
const SAMPLE_RATE: usize = 16_000;

pub const N_FRAMES: usize = 3_000;
const N_FFT: usize = 400;
const HOP_LENGTH: usize = 160;
const TARGET_SAMPLES: usize = 30 * SAMPLE_RATE;

pub fn pad_or_truncate_to_30s(audio: &[f32]) -> Vec<f32> {
    let mut out = vec![0.0_f32; TARGET_SAMPLES];
    let n = audio.len().min(TARGET_SAMPLES);
    out[..n].copy_from_slice(&audio[..n]);
    out
}

pub struct WhisperMel {
    pub n_mels: usize,
    filters: Vec<f32>,
    hann: Vec<f32>,

    fft: std::sync::Arc<dyn realfft::RealToComplex<f32>>,
}

impl WhisperMel {
    pub fn new(n_mels: usize) -> Self {
        let mut planner = realfft::RealFftPlanner::<f32>::new();
        let fft = planner.plan_fft_forward(N_FFT);
        Self {
            n_mels,
            filters: crate::mel_scale::build_mel_filters(n_mels, N_FFT, SAMPLE_RATE),
            hann: crate::mel_scale::hann_window(N_FFT),
            fft,
        }
    }

    pub fn log_mel(&self, audio_30s: &[f32]) -> Vec<f32> {
        assert_eq!(audio_30s.len(), TARGET_SAMPLES, "expected 30s of audio");
        let n_bins = N_FFT / 2 + 1;

        let mut input_buf = self.fft.make_input_vec();
        let mut output_buf = self.fft.make_output_vec();

        let pad = N_FFT / 2;
        let mut padded = vec![0.0_f32; audio_30s.len() + N_FFT];
        for i in 0..pad {
            padded[i] = audio_30s[pad - i];
        }
        padded[pad..pad + audio_30s.len()].copy_from_slice(audio_30s);
        for i in 0..pad {
            let src = audio_30s.len().saturating_sub(2 + i);
            padded[pad + audio_30s.len() + i] = audio_30s[src];
        }

        let mut mel = vec![0.0_f32; self.n_mels * N_FRAMES];
        let mut power = vec![0.0_f32; n_bins];
        for frame in 0..N_FRAMES {
            let start = frame * HOP_LENGTH;
            for i in 0..N_FFT {
                input_buf[i] = padded[start + i] * self.hann[i];
            }
            self.fft
                .process(&mut input_buf, &mut output_buf)
                .expect("FFT process");
            for (k, c) in output_buf.iter().enumerate() {
                power[k] = c.re * c.re + c.im * c.im;
            }
            for m in 0..self.n_mels {
                let row = &self.filters[m * n_bins..(m + 1) * n_bins];
                let mut sum = 0.0_f32;
                for k in 0..n_bins {
                    sum += row[k] * power[k];
                }
                mel[m * N_FRAMES + frame] = sum;
            }
        }

        let eps = 1e-10_f32;
        let mut max_val = f32::NEG_INFINITY;
        for v in mel.iter_mut() {
            *v = v.max(eps).log10();
            if *v > max_val {
                max_val = *v;
            }
        }
        let floor = max_val - 8.0;
        for v in mel.iter_mut() {
            if *v < floor {
                *v = floor;
            }
            *v = (*v + 4.0) / 4.0;
        }
        mel
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pad_short_audio() {
        let p = pad_or_truncate_to_30s(&[1.0, 2.0, 3.0]);
        assert_eq!(p.len(), TARGET_SAMPLES);
        assert_eq!(&p[..3], &[1.0, 2.0, 3.0]);
        assert_eq!(p[3], 0.0);
    }

    #[test]
    fn truncate_long_audio() {
        let big = vec![1.0_f32; TARGET_SAMPLES + 100];
        let p = pad_or_truncate_to_30s(&big);
        assert_eq!(p.len(), TARGET_SAMPLES);
    }

    #[test]
    fn mel_shape_for_80() {
        let m = WhisperMel::new(80);
        let audio = vec![0.1_f32; TARGET_SAMPLES];
        let mel = m.log_mel(&audio);
        assert_eq!(mel.len(), 80 * N_FRAMES);
        for v in &mel {
            assert!(v.is_finite());
        }
    }

    #[test]
    fn mel_shape_for_128() {
        let m = WhisperMel::new(128);
        let audio = vec![0.0_f32; TARGET_SAMPLES];
        let mel = m.log_mel(&audio);
        assert_eq!(mel.len(), 128 * N_FRAMES);
    }
}
